docs/kb/laptop-dev-env.md

```diff
@@ -0,0 +1,224 @@
+of the Embedded Controller.
+
+| Size | 14 inch models | 16 inch models |
+| --- | --- | --- |
+| Programmer | Arduino Mega 2560 | Arduino Mega 2560 |
+| Cable | **FFC 0.5mm pitch 24 pin** | **FFC 1.0mm pitch 24 pin** |
+| Software | ecflash | ecflash |
+
+#### Preparation
+
+1. Obtain a copy of the EC firmware and tools:
+
+    ```bash
+    git clone https://github.com/Dasharo/ec.git
+    cd ec
+    git submodule update --init --checkout
+    ./scripts/deps.sh
+    ```
+
+1. Follow the on-screen instructions to finish configuring your development
+environment.
+1. Connect the Arduino to your workstation with a USB-A to USB-B cable. Do not
+connect the target laptop yet.
+1. Flash the Arduino with debugger firmware:
+
+    ```bash
+    make BOARD=arduino/mega2560
+    make BOARD=arduino/mega2560 flash
+    ```
+
+#### Flash the target laptop
+
+1. In the EC directory, build the EC flashing application:
+
+    ```bash
+    cargo build --manifest-path ecflash/Cargo.toml --example isp --release
+    ```
+
+1. Connect the Arduino to the target laptop with the appropriate FFC cable for
+   your model. If the cable is narrower than the socket on the laptop, align the
+   cable and socket to pin 0 (to the left).
+1. Connect a second cable (e.g. USB-C to USB-C) from the workstation to the
+   target - this cable is needed provide grounding.
+1. Flash the target:
+
+    ```bash
+    sudo ecflash/target/release/examples/isp [path/to/firmware.rom]
+    ```
+
+##### Troubleshooting
+
+In case flashing fails, try the following:
+
+- Try running the command again, sometimes it starts working on the third
+  attempt
+- Try another grounding cable, USB-A to USB-C, or even HDMI to HDMI have been
+  shown to work
+- Try with / without the internal battery connected
+- Try with / without the AC adapter connected
+- Try plugging in the AC adapter right before / after running the command (last
+  resort)
+
+### Building
+
+Build EC firmware:
+
+```bash
+make BOARD=your/board
+```
+
+### Debugging
+
+#### Build EC firmware with debugger support enabled
+
+1. In file `src/board/system76/common/common.mk`, enable parallel port debugging
+    and set the log level to DEBUG:
+
+    ```makefile
+    # Set log level
+    # 0 - NONE
+    # 1 - ERROR
+    # 2 - WARN
+    # 3 - INFO
+    # 4 - DEBUG
+    # 5 - TRACE
+    CFLAGS+=-DLEVEL=4
+
+    # Uncomment to enable debug logging over keyboard parallel port
+    CFLAGS+=-DPARALLEL_DEBUG
+
+    # Uncomment to enable I2C debug on 0x76
+    #CFLAGS+=-DI2C_DEBUGGER=0x76
+    ```
+
+1. Build firmware for the target:
+
+    ```bash
+    make BOARD=your/board
+    ```
+
+1. Flash the firmware to the target:
+
+    ```bash
+    make BOARD=your/board flash_external
+    # or
+    sudo ecflash/target/release/examples/isp path/to/your/ec.rom
+    ```
+
+1. Start the console:
+
+    ```bash
+    make BOARD=your/board console_external
+    ```
+
+1. Power on the target and observe the console for logs.
```

Comment:
Hmm AFAIK, the flashing has to be done on disconnected battery. The logs start to appear once AC is plugged.

Reply:
Kind of. The logs will be printed if the EC finds the debugger while powering on, so theoretically holding the power button for 10 seconds to reset the EC will work too. But you're right, this should be noted here.

Reply:
If you update this section, let me know.

Reply:
6cac4ed62ff5ffcfc40e48cf691852b91aa29dbe